readme.md
# Bolivar
> Federated Open-Registry proxy for p2p sharing of packages


[![Imagelink](https://i.imgur.com/g9jEN6T.png)](https://www.youtube.com/watch?v=VA1HibxXOx0)

## Introduction

Bolivar is a prototype to move the registry closer to wherever you are, and
enable local network sharing of already downloaded data.

This means:

- Faster performance as mirrors are closer to you
- You can easily install a registry proxy in your local network
- You'll be able to install modules from your friends next to you, even
  without a upstream internet connection

## Usage

Install bolivar.

Then you just run `bolivar` and it's up and running!

Next step is pointing your package manager to it:

```bash
# replace npm with yarn or pnpm if you use those
npm config set registry http://localhost:8080
```

You can see all the different options by doing `bolivar --help`

```
$ bolivar --help
usage: bolivar-vlatest-linux-amd64 [<flags>]

A federated version of Open-Registry (https://open-registry.dev)

Flags:
      --help                Show context-sensitive help (also try --help-long and --help-man).
      --federate="/dns4/npm.open-registry.dev/tcp/4001/ipfs/QmbNjMCXkwt7fyBBWc3R8mZzrX8KebkCo4Qv67wVmCH5Aa"
                            Multiaddr of primary federation server to read from
  -s, --share               Whetever to share downloaded packages to others
      --update-type="http"  How to receive updates to root hash.
  -o, --offline             Disable any external connections
  -v, --verbose             Verbose mode.
  -l, --listen-addr="/ip4/0.0.0.0/tcp/4005"
                            What address to set the libp2p node to listen at
  -r, --repo-path="ipfs-test-repo"
                            What path to use for the repository
  -a, --http-address="localhost"
                            What address to use for HTTP server
  -p, --http-port="8080"    What port to use for HTTP server
      --http-endpoint="https://npm.open-registry.dev"
                            What the endpoint for registry index is
      --version             Show application version.
```

Example output when running:

```
$ bolivar
Starting libp2p host
libp2p listening on the following addresses:
/ip4/127.0.0.1/tcp/4005
/ip4/192.168.2.238/tcp/4005
/ip4/172.17.0.1/tcp/4005
Connecting to open-registry.dev libp2p node...
Connected
Think we protected a bootstrap peer from being killed too
Starting server localhost:8080...
0 B|0 B [0 B/s|0 B/s] (Peers: 1)
19 kB|832 B [19 kB/s|831 B/s] (Peers: 1)
19 kB|832 B [7.0 kB/s|306 B/s] (Peers: 1)
19 kB|832 B [2.6 kB/s|112 B/s] (Peers: 1)
19 kB|832 B [942 B/s|41 B/s] (Peers: 1)
19 kB|832 B [346 B/s|15 B/s] (Peers: 3)
```

## Contributing

### Testing

- `make test` - runs bolivar and tries to download dependencies for webpack/webpack via bolivar

### Building

- `make linux` / `make windows` / `make darwin` builds binary for linux/windows/darwin

### Linting

- `make lint` runs gometalinter over all the .go source files in the repo

## License

MIT 2019 - Open-Registry


CHANGELOG.md
### 0.1.0 - 2019-05-18

Initial version! Everything changes if you compare to the previous version.


test.sh
#! /usr/bin/env bash

set -ex

docker build -t open-services/bolivar .

CONTAINER_ID=$(docker run -d open-services/bolivar)

function finish {
  echo "Showing you the logs from the container before kill/rm"
  docker logs $CONTAINER_ID
  echo "## END LOGS"
  docker kill $CONTAINER_ID
  docker rm $CONTAINER_ID
}
trap finish EXIT

echo $CONTAINER_ID

docker exec -it $CONTAINER_ID git clone https://github.com/webpack/webpack.git --depth=1 /app

docker exec -it $CONTAINER_ID yarn config set registry http://localhost:8080

# Give the Bolivar HTTP server a chance to start
sleep 5

time docker exec -it $CONTAINER_ID bash -c "cd /app && rm yarn.lock && yarn install --verbose --non-interactive"


main.go
package main

import (
	"fmt"
	"log"
	"time"

	ipfslite "github.com/hsanjuan/ipfs-lite"
	host "github.com/libp2p/go-libp2p-host"
	multiaddr "github.com/multiformats/go-multiaddr"
	"github.com/open-services/bolivar/cli"
	http "github.com/open-services/bolivar/http"
	"github.com/open-services/bolivar/p2p"
	"github.com/open-services/bolivar/update"
)

// split to packages
var appConfig cli.Config

// What the version of the application is
var appVersion = "0.1.0"

// What the author of the application is
var appAuthor = "Open-Registry"

// libp2p peer to use for transfer
var libp2pNode *ipfslite.Peer

func main() {

	appConfig = cli.Init(appVersion, appAuthor)

	if appConfig.Verbose {
		cli.PrintConfig(appConfig)
	}

	updater := update.NewUpdater(appConfig)

	go func() {
		for {
			time.Sleep(1 * time.Second)
			http.SetRootHash(updater.Update())
		}
	}()
	http.SetRootHash(updater.Update())

	libp2p, h := p2p.StartLibp2p(appConfig)
	libp2pNode = libp2p

	addrs := h.Addrs()
	fmt.Println("libp2p listening on the following addresses:")
	for _, addr := range addrs {
		fmt.Println(addr)
	}

	connectAndReconnect(libp2pNode, h, appConfig.FederateAddr)

	// TODO should start immediately and hang requests until we have the libp2p
	// node ready and connected
	log.Fatal(http.StartServer(appConfig, libp2pNode))
}

// Ensures we stay connected to this peer.
// In case where peer got disconnected, try to reconnect
func connectAndReconnect(libp2pNode *ipfslite.Peer, host host.Host, ma string) {
	err := p2p.ConnectToPeer(libp2pNode, host, ma)
	if err != nil {
		panic(err)
	}
	// make sure we're connected to federate addr always
	go func() {
		fedID, err := getPeerIDFromMultiaddr(ma)
		if err != nil {
			panic(err)
		}
		for {
			time.Sleep(5 * time.Second)
			conns := host.Network().Conns()

			foundOR := false
			for _, conn := range conns {
				if conn.RemotePeer().String() == fedID {
					foundOR = true
				}
			}

			if !foundOR {
				log.Println("Lost connection to OR for some reason")
				log.Println("reconnecting")
				err := p2p.ConnectToPeer(libp2pNode, host, appConfig.FederateAddr)
				if err != nil {
					panic(err)
				}
				log.Println("reconnected")
			}
		}
	}()

}

func getPeerIDFromMultiaddr(maStr string) (string, error) {
	ma, err := multiaddr.NewMultiaddr(maStr)
	if err != nil {
		return "", err
	}
	// Protocol ID from https://github.com/multiformats/go-multiaddr/blob/e1825f7b50d1dcebdaa28bc31a310fa2be4c00ee/protocols.go#L17

	fedID, err := ma.ValueForProtocol(0x01A5)
	if err != nil {
		return "", err
	}
	return fedID, nil
}


p2p/p2p.go
package p2p

import (
	"context"
	"fmt"
	"strconv"
	"time"

	humanize "github.com/dustin/go-humanize"
	ipfslite "github.com/hsanjuan/ipfs-lite"
	ds "github.com/ipfs/go-datastore"
	ds_sync "github.com/ipfs/go-datastore/sync"
	golog "github.com/ipfs/go-log"
	libp2p "github.com/libp2p/go-libp2p"
	connmgr "github.com/libp2p/go-libp2p-connmgr"
	crypto "github.com/libp2p/go-libp2p-crypto"
	host "github.com/libp2p/go-libp2p-host"
	ipnet "github.com/libp2p/go-libp2p-interface-pnet"
	dht "github.com/libp2p/go-libp2p-kad-dht"
	metrics "github.com/libp2p/go-libp2p-metrics"
	peerstore "github.com/libp2p/go-libp2p-peerstore"
	pnet "github.com/libp2p/go-libp2p-pnet"
	"github.com/libp2p/go-libp2p/p2p/discovery"
	routedhost "github.com/libp2p/go-libp2p/p2p/host/routed"
	multiaddr "github.com/multiformats/go-multiaddr"
	"github.com/open-services/bolivar/cli"
)

// CurrentStats is the latest received metrics for current/total bandwidth
var CurrentStats *metrics.Stats

func setupConnMgr() *connmgr.BasicConnMgr {
	cmgr := connmgr.NewConnManager(10, 50, 1*time.Second)
	go func() {
		nctx, cancel := context.WithCancel(context.Background())
		for {
			time.Sleep(10 * time.Second)
			if nctx == nil {
				nctx, cancel = context.WithCancel(context.Background())
			}
			cmgr.TrimOpenConns(nctx)
			cancel()
			cancel = nil
			nctx = nil
		}
	}()
	return cmgr
}

// SetupLibp2p ...
func SetupLibp2p(
	ctx context.Context,
	hostKey crypto.PrivKey,
	secret []byte,
	listenAddrs []multiaddr.Multiaddr,
) (host.Host, *dht.IpfsDHT, error) {

	var prot ipnet.Protector
	var err error

	// Create protector if we have a secret.
	if len(secret) > 0 {
		var key [32]byte
		copy(key[:], secret)
		prot, err = pnet.NewV1ProtectorFromBytes(&key)
		if err != nil {
			return nil, nil, err
		}
	}

	rep := metrics.NewBandwidthCounter()

	h, err := libp2p.New(
		ctx,
		libp2p.Identity(hostKey),
		libp2p.ListenAddrs(listenAddrs...),
		libp2p.PrivateNetwork(prot),
		libp2p.NATPortMap(),
		libp2p.ConnectionManager(setupConnMgr()),
		libp2p.BandwidthReporter(rep),
	)

	if err != nil {
		return nil, nil, err
	}

	firstResult := false

	// reporting
	go func() {
		for {
			time.Sleep(100 * time.Millisecond)
			// bandwidth
			stats := rep.GetBandwidthTotals()
			CurrentStats = &stats
			firstResult = true
		}
	}()

	go func() {
		for {
			time.Sleep(1 * time.Second)
			// peers
			conns := h.Network().Conns()
			numPeers := strconv.Itoa(len(conns))

			if firstResult {
				stats := CurrentStats

				// spew.Dump(stats)
				bwMessage := "%s|%s [%s/s|%s/s] (Peers: %s)"
				fmt.Println(fmt.Sprintf(bwMessage,
					humanize.Bytes(uint64(stats.TotalIn)),
					humanize.Bytes(uint64(stats.TotalOut)),
					humanize.Bytes(uint64(stats.RateIn)),
					humanize.Bytes(uint64(stats.RateOut)),
					numPeers,
				))
			}
		}
	}()

	dstore := ds_sync.MutexWrap(ds.NewMapDatastore())
	idht := dht.NewDHTClient(ctx, h, dstore)
	rHost := routedhost.Wrap(h, idht)
	return rHost, idht, nil
}

type discoveryHandler struct {
	host host.Host
}

const discoveryConnTimeout = time.Second * 30

func (dh *discoveryHandler) HandlePeerFound(p peerstore.PeerInfo) {
	// fmt.Println("found local peer")
	// // log.Warning("trying peer info: ", p)
	// spew.Dump(p.Addrs)
	ctx, cancel := context.WithTimeout(context.Background(), discoveryConnTimeout)
	defer cancel()
	// important connection!
	if err := dh.host.Connect(ctx, p); err != nil {
		fmt.Println("Failed to connect to peer found by discovery: ", err)
	}
	dh.host.ConnManager().Protect(p.ID, "local")
	fmt.Println("Think we protected a bootstrap peer from being killed too")
}

// StartLibp2p ...
// TODO needs to have a connmgr as we're connecting to all peers currently
// TODO also include bandwidth reporter for nicer ui
func StartLibp2p(config cli.Config) (*ipfslite.Peer, host.Host) {
	fmt.Println("Starting libp2p host")
	ctx := context.Background()

	err := golog.SetLogLevel("*", "notice")
	// err := golog.SetLogLevel("*", "info")
	if err != nil {
		panic(err)
	}

	ds, err := ipfslite.BadgerDatastore(config.RepoPath)
	if err != nil {
		panic(err)
	}
	priv, _, err := crypto.GenerateKeyPair(crypto.RSA, 2048)
	if err != nil {
		panic(err)
	}

	// TODO configurable
	listen, _ := multiaddr.NewMultiaddr(config.ListenAddr)

	h, dht, err := SetupLibp2p(
		ctx,
		priv,
		nil,
		[]multiaddr.Multiaddr{listen},
	)

	if err != nil {
		panic(err)
	}

	srv, err := discovery.NewMdnsService(context.Background(), h, 5*time.Second, discovery.ServiceTag)

	if err != nil {
		panic(err)
	}

	n := &discoveryHandler{h}

	srv.RegisterNotifee(n)

	lite, err := ipfslite.New(ctx, ds, h, dht, nil)
	if err != nil {
		panic(err)
	}
	return lite, h
}

// ConnectToPeer bootstraps with a peer to keep the connection open for as long as possible
func ConnectToPeer(node *ipfslite.Peer, h host.Host, ma string) error {
	fmt.Println("Connecting to open-registry.dev libp2p node...")
	openRegistryAddr, err := multiaddr.NewMultiaddr(ma)
	if err != nil {
		return err
	}
	orPeerInfo, err := peerstore.InfoFromP2pAddr(openRegistryAddr)
	if err != nil {
		return err
	}
	peers := []peerstore.PeerInfo{}
	peers = append(peers, *orPeerInfo)
	// mark as important so connmgr doesn't kill connection
	node.Bootstrap(peers)
	fmt.Println("Connected")
	h.ConnManager().Protect(orPeerInfo.ID, "bootstrap")
	fmt.Println("Think we protected a bootstrap peer from being killed too")
	return nil
}


update/http.go
package update

import (
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/open-services/bolivar/cli"
)

// HTTPUpdater receives root hash updates via polling a HTTP endpoint
type HTTPUpdater struct {
	Config cli.Config
}

type httpRes struct {
	Hash           string `json:"hash"`
	CumulativeSize int    `json:"cumulativesize"`
	Blocks         int    `json:"blocks"`
}

// helper function to receive json from endpoint
func getJSON(userAgent, url string, target interface{}) error {
	r, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}
	r.Header.Set("User-Agent", userAgent)
	resp, err := myClient.Do(r)
	if err != nil {
		return err
	}
	defer func() {
		err := resp.Body.Close()
		if err != nil {
			panic(err)
		}
	}()

	return json.NewDecoder(resp.Body).Decode(target)
}

// NewHTTPUpdater returns a new http updater...
func NewHTTPUpdater(config cli.Config) *HTTPUpdater {
	hu := new(HTTPUpdater)
	hu.Config = config
	return hu
}

var myClient = &http.Client{Timeout: 10 * time.Second}

// Update receives latest root hash from open-registry
// TODO write to repo path so the next time we start it, we have a hash
// to go by initially
func (hu *HTTPUpdater) Update() string {
	endpoint := hu.Config.HTTPEndpoint
	isVerbose := hu.Config.Verbose

	if isVerbose {
		log.Println("Fetching latest root hash from " + endpoint)
	}

	userAgent := "bolivar/0.1.0"

	res := new(httpRes)
	err := getJSON(userAgent, endpoint, res)

	if err != nil {
		log.Println("Seems there are issues contacting to " + endpoint + " right now")
		if isVerbose {
			log.Println(err)
		}
	}

	if isVerbose {
		log.Println("Fetched, latest hash is " + res.Hash)
	}

	return res.Hash
}


update/update.go
package update

import (
	"log"

	"github.com/open-services/bolivar/cli"
)

// Updater is a mechanism for receiving a new root hash
type Updater interface {
	// Start(interval time.Duration)
	Update() string
}

// checks if a string in a list of strings exists
// (how is this not in the standard library???)
func stringInSlice(a string, list []string) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}

// valid update types
var updateTypes = []string{"http", "dns", "ipns", "pubsub"}

// NewUpdater returns a updater based on the `update-type` in config
func NewUpdater(config cli.Config) Updater {
	updateType := config.UpdateType
	if stringInSlice(updateType, updateTypes) {
		// Currently only supports http
		if updateType != "http" {
			log.Fatal("Update type " + updateType + " is not yet supported")
		}
	} else {
		log.Fatal("Update type " + updateType + " is not supported")
	}

	updater := NewHTTPUpdater(config)
	return updater
}


simulator/src/simulator/core.clj
(ns simulator.core
  (:require [clj-docker-client.core :as docker]
            [clojure.test :refer [is]]
            [cheshire.core :refer [parse-string]]
            [clojure.contrib.humanize :refer [filesize]])
  (:gen-class))

;; (comment
;;   ;; start five registries with share off
;;   ;; download packages
;;   ;; start five registries with share on
;;   ;; download packages
;;   ;; second version is faster
;; )
;; 
;; ;; faster as it's using sharing
;; (def open-registry-id (run-open-registry))
;; (def bolivar-ids (map (partial run bolivar) (range 5)))
;; 
;; ;; todo
;; ;; - [ ] image for open-registry
;; ;; - [ ] rename open-registry-fed -> bolivar
;; ;; - [ ] methods for
;; ;;  - [ ] running open-registry
;; ;;  - [ ] running bolivar
;; ;;  - [ ] running commands in existing container

(def open-registry {:image "open-services/open-registry:latest"})

(def bolivar {:image "open-services/bolivar:latest"
              :cmd "bolivar --http-address=0.0.0.0"})

(def deps-test {:image "open-services/deps-test:latest"})

(def conn (docker/connect))

(defn run [app]
  (clojure.pprint/pprint app)
  (docker/start conn (docker/create conn (:image app) (:cmd app) {} (:ports app))))

(defn exec [container-id cmd]
  (println container-id)
  (println cmd))

;; Stolen from
;; https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core.clj#L3850
(defmacro measure
  "Evaluates expr and return the time (in milliseconds) it took together with the result"
  [expr]
  `(let [start# (. System (nanoTime))
         ret# ~expr
         duration# (/ (double (- (. System (nanoTime)) start#)) 1000000.0)]
     {:return ret#
      :duration duration#}))

(comment
  (measure (+ 1 1))
  (measure (Thread/sleep 1000))
  )

(defn run-tests [tests]
  (doseq [t tests]
    (let [ret (t)]
      (clojure.pprint/pprint ret))))

(defn get-ip [container-id]
  (-> (docker/inspect conn container-id)
      :NetworkSettings
      :Networks
      :bridge
      :IPAddress
      ))

(defn run-with-registry [ip]
  (format "bash -c \"yarn --verbose --registry=http://%s:8080\"" ip))

(defn deps-test-image [ip]
  {:image "open-services/deps-test:latest"
   :cmd (run-with-registry ip)})

(defn run-and-wait [conn image]
  (let [id (run image)]
    (docker/wait-container conn id)
    id))

(defn measure-install [conn bolivar-ip]
  (measure (run-and-wait conn (deps-test-image bolivar-ip))))

(defn bolivar-metrics [bolivar-ip]
  (slurp (format "http://%s:8080/_api/metrics" bolivar-ip)))

(defn clean [ids]
  (doseq [id ids]
    (try (docker/kill conn id)
         (catch Exception e (println (str "didnt find container " id))))
    (docker/rm conn id)))

(defn get-metric [bolivar-ip k]
  (k (parse-string (bolivar-metrics bolivar-ip) true)))

(defn humanize-metrics [bolivar-ip]
  (let [m(parse-string (bolivar-metrics bolivar-ip) true)]
  {:TotalIn (filesize (:TotalIn m))
   :TotalOut (filesize (:TotalOut m))
   :RateIn (filesize (:RateIn m))
   :RateOut (filesize (:RateOut m))}))

;; (clojure.pprint/pprint (humanize-metrics bolivar-ip))

;; test cases

(defn second-install-cached []
  (let [bolivar-id (run bolivar)
        bolivar-ip (get-ip bolivar-id)]
    (Thread/sleep 1000)
    (let [m1 (measure-install conn bolivar-ip)
          m2 (measure-install conn bolivar-ip)]
      (is (< (:duration m2) (:duration m1)))
      (clean [(:return m2) (:return m1)
              bolivar-id]))))

(defn second-bolivar-faster []
  (let [bolivar-id (run bolivar)
        bolivar-ip (get-ip bolivar-id)
        bolivar2-id (run bolivar)
        bolivar2-ip (get-ip bolivar2-id)]
    (Thread/sleep 1000)
    (let [m1 (measure-install conn bolivar-ip)
          m2 (measure-install conn bolivar2-ip)]
      (let [ti1 (get-metric bolivar-ip :TotalIn)
            ti2 (get-metric bolivar2-ip :TotalIn)
            to1 (get-metric bolivar-ip :TotalOut)
            to2 (get-metric bolivar2-ip :TotalOut)]
        (is (< (:duration m2) (:duration m1)))
        (is (> to1 ti2))
        (is (> to1 to2))
        (is (> ti1 0))
        (is (> ti2 0))
        (is (> to1 0))
        (is (> to2 0))
        (clean [(:return m2) (:return m1)
                bolivar-id bolivar2-id])))))

(defn -main
  [& args]
  (run-tests [second-install-cached
              second-bolivar-faster])
  (System/exit 0))


cli/cli.go
package cli

import (
	"log"
	"path"

	"github.com/davecgh/go-spew/spew"
	homedir "github.com/mitchellh/go-homedir"
	kingpin "gopkg.in/alecthomas/kingpin.v2"
)

// TODO should be using dnsaddr (currently hardcoded peerID)
var openRegistryMultiaddr = "/dns4/npm.open-registry.dev/tcp/4001/ipfs/QmdAgn46prB29nfW1fSoKJsehma7DL6C3yygjCaVKYxcFw"

// what path should our libp2p node listen at
var defaultListenAddr = "/ip4/0.0.0.0/tcp/4005"

// what path should we use to store data
// TODO should point to directory in homedir by default
var defaultRepoPath = getDefaultRepoPath()

// what address to set HTTP server to listen to
var defaultHTTPAddress = "localhost"

// what port to set HTTP server to listen to
var defaultHTTPPort = "8080"

// default http update endpoint
var defaultHTTPUpdateEndpoint = "https://npm.open-registry.dev"

// cli args
var (
	federateAddr = kingpin.Flag("federate", "Multiaddr of primary federation server to read from").Default(openRegistryMultiaddr).String()
	share        = kingpin.Flag("share", "Whetever to share downloaded packages to others").Default("true").Short('s').Bool()
	updateType   = kingpin.Flag("update-type", "How to receive updates to root hash.").Default("http").String()
	offline      = kingpin.Flag("offline", "Disable any external connections").Short('o').Bool()
	verbose      = kingpin.Flag("verbose", "Verbose mode.").Short('v').Bool()
	listenAddr   = kingpin.Flag("listen-addr", "What address to set the libp2p node to listen at").Default(defaultListenAddr).Short('l').String()
	repoPath     = kingpin.Flag("repo-path", "What path to use for the repository").Default(defaultRepoPath).Short('r').String()
	httpAddress  = kingpin.Flag("http-address", "What address to use for HTTP server").Default(defaultHTTPAddress).Short('a').String()
	httpPort     = kingpin.Flag("http-port", "What port to use for HTTP server").Default(defaultHTTPPort).OverrideDefaultFromEnvar("PORT").Short('p').String()
	httpEndpoint = kingpin.Flag("http-endpoint", "What the endpoint for registry index is").Default(defaultHTTPUpdateEndpoint).String()
)

// Config holds all the configuration variables from parsing the CLI args and flags
type Config struct {
	Offline      bool
	Verbose      bool
	Share        bool
	FederateAddr string
	UpdateType   string
	ListenAddr   string
	RepoPath     string
	HTTPAddress  string
	HTTPPort     string
	HTTPEndpoint string
}

func getDefaultRepoPath() string {
	home, err := homedir.Dir()
	if err != nil {
		panic(err)
	}
	p := path.Join(home, ".bolivar")
	return p
}

// Init initializes the cli
func Init(version, author string) Config {
	kingpin.UsageTemplate(kingpin.CompactUsageTemplate).Version(version).Author(author)
	kingpin.CommandLine.Help = "A federated version of Open-Registry (https://open-registry.dev)"
	kingpin.Parse()

	cfg := Config{
		FederateAddr: *federateAddr,
		Share:        *share,
		UpdateType:   *updateType,
		Offline:      *offline,
		Verbose:      *verbose,
		ListenAddr:   *listenAddr,
		RepoPath:     *repoPath,
		HTTPAddress:  *httpAddress,
		HTTPPort:     *httpPort,
		HTTPEndpoint: *httpEndpoint,
	}

	return cfg
}

// PrintConfig is a helper function to print out all config options
func PrintConfig(config Config) {
	log.Println("CLI flags and args:")
	spew.Dump(config)
}
